### PR TITLE
Toggle windows on keybind press

### PR DESCRIPTION
**Pull request by LeftofZen**
Implements #1626 
Currently just proof-of-concept for map window, but I think it works great and feels intuitive.

### CODE REVIEW

**Comment by LeftofZen** [comment]
@AaronVanGeffen I see you added the discussion tag - I've implemented this feature fully so feel free to check it out and test how it see feels in game. The one thing I couldn't figure out was how to toggle the same window but different modes, eg the `construct airport` vs `construct shipport` windows are indistinguishable at the 'toggleWindow' call site.

File: src/OpenLoco/Input/ShortcutManager.cpp [hunk 1]
```diff
@@ -299,13 +299,41 @@ namespace OpenLoco::Input::ShortcutManager
...
+            WindowManager::close(window);
+        }
+        else
+        {
+            WindowManager::bringToFront(window);
+        }
+    }
+
+    template<typename TFunc, typename... TFuncArgs>
+    static void toggleWindow(WindowType windowType, TFunc windowOpener, TFuncArgs... args)
```
Comment (LeftofZen) [inline src/OpenLoco/Input/ShortcutManager.cpp:325]: helper method so we don't have to call std::bind at every call site

File: src/OpenLoco/Input/ShortcutManager.cpp [hunk 2]
```diff
@@ -455,7 +484,7 @@ namespace OpenLoco::Input::ShortcutManager
         if (isEditorMode())
             return;
 
-        Ui::Windows::CompanyWindow::open(CompanyManager::getControllingId());
+        toggleWindow(WindowType::company, Windows::CompanyWindow::open, CompanyManager::getControllingId());
```
Comment (duncanspumpkin) [inline src/OpenLoco/Input/ShortcutManager.cpp:487]: Does this actually work. You will probably end up closing an unrelated companies window with this button. Say i had 3 company windows open. My understanding is that 3 presses of this shortcut would close the 3 windows and not open three players company window until the 4th press.

**Comment by duncanspumpkin** [comment]
I think you'll need to have 2 versions of the helper toggle code. One for windows that are singular and one for windows that take a window number (e.g. company window) and there can be multiple. I can't actually remember if the company window can have multiple open so perhaps I'm over thinking this.

File: src/OpenLoco/Input/ShortcutManager.cpp (hunk 1 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Input/ShortcutManager.cpp:325]: Couldn't you just write lambdas instead of messing with a bind?

**Comment by LeftofZen** [comment]
Hmm ok I'll take a look at the company windows more, I didn't see what happens with multiple open

File: src/OpenLoco/Input/ShortcutManager.cpp (hunk 1 above)
Comment (LeftofZen) [reply src/OpenLoco/Input/ShortcutManager.cpp:325]: Yes, but that would have been more code. I can do the lambdas if you prefer, but I think its much cleaner this way

**Comment by LeftofZen** [comment]
You're right, the company windows don't play too nicely with this (ie they don't toggle as expected), I'll see if there's anything I can do to improve it

**Comment by LeftofZen** [comment]
Was thinking about this one, for the windows with some funky behaviour (ie company window, some of the build windows with tabs) I could just not add the 'toggle' behaviour for now, and keep it on the simple windows that do work with it.

**Comment by duncanspumpkin** [comment]
Yes that would be a good idea just remove the problem ones for now.

**Comment by LeftofZen** [comment]
Won't be merging this for 22.10, would like to improve the behaviour with tabbed windows first

**Pull request closed without merging**